File: mcomix/layout.py
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

from mcomix.enums import Scroll
from mcomix.hyperrectangles import Box


class FiniteLayout:  # 2D only
    def __init__(self, content_sizes: list[tuple], viewport_size: tuple, orientation: list,
                 distribution_axis: int, alignment_axis: int):
        """
        Lays out a finite number of Boxes along the first axis.

        :param content_sizes: The sizes of the Boxes to lay out.
        :param viewport_size: The size of the viewport.
        :param orientation: The orientation to use.
        :param distribution_axis: the axis along which the Boxes are distributed.
        :param alignment_axis: the axis to center
        """

        super().__init__()

        # reverse order if necessary
        if orientation[distribution_axis] == -1:
            content_sizes = tuple(reversed(content_sizes))
        temp_cb_list = tuple(map(Box, content_sizes))

        # align to center
        temp_cb_list = Box.align_center(temp_cb_list, alignment_axis, 0, orientation[alignment_axis])

        # distribute
        temp_cb_list = Box.distribute(temp_cb_list, distribution_axis, 0)
        temp_bb = Box.bounding_box(temp_cb_list).wrapper_box(viewport_size, orientation)

        # move to global origin
        bbp = temp_bb.get_position()
        for idx, item in enumerate(temp_cb_list):
            temp_cb_list[idx] = temp_cb_list[idx].translate_opposite(bbp)

        temp_bb = temp_bb.translate_opposite(bbp)
        # reverse order again, if necessary
        if orientation[distribution_axis] == -1:
            temp_cb_list = tuple(reversed(temp_cb_list))

        # done
        self.__content_boxes = temp_cb_list
        self.__union_box = temp_bb
        self.__viewport_box = Box(viewport_size)
        self.__orientation = orientation

    def scroll_to_predefined(self, destination: tuple):
        """
        Returns a new viewport position when scrolling towards a
        predefined destination. Note that all params are lists of integers
        where each index corresponds to one dimension.

        :param destination: An integer representing a predefined destination.
        Either 1 (towards the greatest possible values in this dimension),
        -1 (towards the smallest value in this dimension), 0 (keep position),
        SCROLL_TO_CENTER (scroll to the center of the content in this
        dimension), SCROLL_TO_START (scroll to where the content starts in this
        dimension) or SCROLL_TO_END (scroll to where the content ends in this dimension).
        :returns: A new viewport position as specified above
        """

        content_position = self.__union_box.get_position()
        content_size = self.__union_box.get_size()
        viewport_size = self.__viewport_box.get_size()
        result = list(self.__viewport_box.get_position())
        for idx, item in enumerate(content_size):
            # The orientation which shows where "forward"
            # points to. Either 1 (towards larger values in this dimension when
            # reading) or -1 (towards smaller values in this dimension when reading).
            o = self.__orientation[idx]
            d = destination[idx]

            if not d:
                continue
            elif d < Scroll.END.value or d > 1:
                raise ValueError(f'invalid destination {d} at index {idx}')
            elif d == Scroll.END.value:
                d = o
            elif d == Scroll.START.value:
                d = -o

            c = content_size[idx]
            v = viewport_size[idx]
            invisible_size = c - v

            if d == Scroll.CENTER.value:
                offset = Box.box_to_center_offset_1d(invisible_size, o)
            else:
                if d == 1:
                    offset = invisible_size
                else:  # d == -1
                    offset = 0

            result[idx] = content_position[idx] + offset

        # Move the viewport to the specified position.
        self.__viewport_box = self.__viewport_box.set_position(tuple(result))

    def get_content_boxes(self):
        """
        Returns the Boxes as they are arranged in this layout.

        :returns: The Boxes as they are arranged in this layout
        """

        return self.__content_boxes

    def get_union_box(self):
        """
        Returns the union Box for this layout.

        :return: The union Box for this layout
        """

        return self.__union_box

    def get_viewport_box(self):
        """
        Returns the current viewport Box.

        :returns: The current viewport Box
        """

        return self.__viewport_box

    def set_orientation(self, orientation: list):
        self.__orientation = orientation

File: mcomix/gui/lib/layout.cpp
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <array>
#include <vector>

#include <ranges>
#include <algorithm>

#include <cassert>

#include "box.hpp"
#include "enums.hpp"
#include "layout.hpp"

Layout::Layout(std::vector<std::array<std::int32_t, 2>> content_sizes, const std::array<std::int32_t, 2>& viewport_size,
               const std::array<std::int32_t, 2>& orientation, const ZoomAxis distribution_axis,
               const ZoomAxis alignment_axis)
{
    // Reverse order if necessary
    if (orientation[static_cast<std::int32_t>(distribution_axis)] == -1)
    {
        std::ranges::reverse(content_sizes);
    }

    for (const auto& size : content_sizes)
    {
        this->content_boxes_.emplace_back(Box({size[0], size[1]}));
    }

    // Align to center
    this->content_boxes_ = Box::align_center(this->content_boxes_,
                                             static_cast<std::int32_t>(alignment_axis),
                                             0,
                                             orientation[static_cast<std::int32_t>(alignment_axis)]);

    // Distribute
    this->content_boxes_ = Box::distribute(this->content_boxes_, static_cast<std::int32_t>(distribution_axis), 0);
    this->union_box_ =
        Box::bounding_box(this->content_boxes_).wrapper_box({viewport_size[0], viewport_size[1]}, orientation);

    // Move to global origin
    const auto bbp = this->union_box_.get_position();
    for (const auto [idx, box] : std::ranges::views::enumerate(this->content_boxes_))
    {
        (void)box;
        this->content_boxes_[idx] = this->content_boxes_[idx].translate_opposite(bbp);
    }

    this->union_box_ = this->union_box_.translate_opposite(bbp);
    // Reverse order again, if necessary
    if (orientation[static_cast<std::int32_t>(distribution_axis)] == -1)
    {
        std::ranges::reverse(this->content_boxes_);
    }

    this->viewport_box_ = Box({viewport_size[0], viewport_size[1]});
    this->orientation_ = orientation;
}

void
Layout::scroll_to_predefined(const std::array<Scroll, 2>& destination)
{
    const auto content_position = this->union_box_.get_position();
    const auto content_size = this->union_box_.get_size();
    const auto viewport_size = this->viewport_box_.get_size();
    auto result = this->viewport_box_.get_position();

    for (std::size_t idx = 0; idx < content_size.size(); ++idx)
    {
        std::int32_t o = this->orientation_[idx];
        std::int32_t d = static_cast<std::int32_t>(destination[idx]);

        if (d == 0)
        {
            continue;
        }

        if (d == static_cast<std::int32_t>(Scroll::END))
        {
            d = static_cast<std::int32_t>(o);
        }
        else if (d == static_cast<std::int32_t>(Scroll::START))
        {
            d = -o;
        }

        std::int32_t c = content_size[idx];
        std::int32_t v = viewport_size[idx];
        std::int32_t invisible_size = c - v;
        std::int32_t offset;

        if (d == static_cast<std::int32_t>(Scroll::CENTER))
        {
            offset = Box::box_to_center_offset_1d(invisible_size, o);
        }
        else
        {
            offset = (d == 1) ? invisible_size : 0;
        }

        result[idx] = content_position[idx] + offset;
    }

    // Move the viewport to the specified position
    this->viewport_box_ = this->viewport_box_.set_position(result);
}

std::vector<Box>
Layout::get_content_boxes() const noexcept
{
    return this->content_boxes_;
}

Box
Layout::get_union_box() const noexcept
{
    return this->union_box_;
}

Box
Layout::get_viewport_box() const noexcept
{
    return this->viewport_box_;
}

const std::array<std::int32_t, 2>
Layout::get_orientation() const noexcept
{
    return this->orientation_;
}

void
Layout::set_orientation(const std::array<std::int32_t, 2>& new_orientation)
{
    this->orientation_ = new_orientation;
}
